src/txdb.cpp

```diff
@@ -566,11 +566,15 @@ bool CBlockTreeDB::LoadBlockIndexGuts(
                     return error("LoadBlockIndex(): CheckProofOfWork failed: %s", pindexNew->ToString());
 
                 // ZIP 221 consistency checks
-                if (chainParams.GetConsensus().NetworkUpgradeActive(pindexNew->nHeight, Consensus::UPGRADE_HEARTWOOD)) {
+                // We assume block index objects on disk that are not at least
+                // CHAIN_HISTORY_ROOT_VERSION were created by nodes that were
+                // not Heartwood aware.
+                if (diskindex.nClientVersion >= CHAIN_HISTORY_ROOT_VERSION &&
+                    chainParams.GetConsensus().NetworkUpgradeActive(pindexNew->nHeight, Consensus::UPGRADE_HEARTWOOD)) {
```

Comment:
I agree that this approach avoids the error triggering. However, it leaves `pindexNew->hashLightClientRoot` potentially incorrect for block index objects that were read with a version less than `CHAIN_HISTORY_ROOT_VERSION`. Perhaps we should fix up `hashLightClientRoot` in that case? I.e. something like

```suggestion
                if (chainParams.GetConsensus().NetworkUpgradeActive(pindexNew->nHeight, Consensus::UPGRADE_HEARTWOOD)) {
                    if (diskindex.nClientVersion < CHAIN_HISTORY_ROOT_VERSION) {
                        pindexNew->hashLightClientRoot = pindexNew->hashChainHistoryRoot;
                    }
```

Reply:
and maybe move the "We assume ..." comment down to line 573.

Reply:
If it was serialized by a client that was not Heartwood-aware, that block header cannot be on a Heartwood-updated chain. Thus, `hashLightClientRoot` will be exactly what's in the block header (which is `hashFinalSaplingRoot`), because that is how `CBlockIndex` parses non-Heartwood aware block indexes.  

`hashChainHistory` will be default set to null (when not explicitly deserialized, which will be the case for these non-Heartwood block headers.)

Reply:
Ah, of course. Yes this is correct as-is.